File: greatestcommondivisorofstrings.js
//Objective is to find the largest substring that divides into both strings.

str1 = "ABCABC", str2 = "ABC"


//O(n) solution that has a pointer on the shorter string and tries every
//possible substring to see if it fits evenly.

//shorter string first
if (str1.length > str2.length) {
    return gcdOfStrings(str2, str1)
}

let start = 1
let length = str2.length
let length2 = str1.length
let max = 0
let result = ''

while (start <= str1.length) {
    let substring = str1.substring(0, start)
    let repeatTimes = length / substring.length
    let repeatTimes2 = length2 / substring.length
    if (substring.repeat(repeatTimes) == str2 && substring.repeat(repeatTimes2) == str1) {
        if (substring.length > max) {
            max = substring.length
            result = substring
        }
    }
    start++
}

return result

File: greatestcommondivisorofstrings.java
//Java Solution

class Solution {
    public String gcdOfStrings(String str1, String str2) {
        if (str1.length() < str2.length()) {
            return gcdOfStrings(str2, str1);
        }
        
        int left = 0;
        int right = 1;
        String ans = "";
        
        while (right <= str1.length()) {
            String substring = str1.substring(left, right++);
            int repeat1 = str1.length() / substring.length();
            int repeat2 = str2.length() / substring.length();
            
            if (substring.repeat(repeat1).equals(str1) && substring.repeat(repeat2).equals(str2)) {
                ans = (ans.length() < substring.length()) ? substring : ans;
            }
        }
        
        return ans;
    }
}